features/crc/mbed_crc.h

```diff
@@ -0,0 +1,223 @@
+        }
+        return CRC_INVALID;
+    }
+
+    /** Compute CRC for the data input
+     *
+     *  @param [IN]  buffer  Data bytes
+     *  @param [IN]  size  Size of data
+     *  @param [OUT] crc  CRC
+     *  @return  0 on success, negative error code on failure
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, crc_size_t *crc) = 0;
+
+    /** Compute partial start, indicate start of partial computation
+     *
+     *  This API should be called before performing any partial computation
+     *  with compute_partial API.
+     *
+     *  @param [OUT] crc  Initial CRC value set by the API
+     *  @return  0  on success or a negative when not supported
+     *  @note: CRC is an out parameter and must be reused with compute_partial
+     *         and compute_partial_stop without any modifications in application.
+     */
+    virtual int32_t compute_partial_start(crc_size_t *crc) const
+    {
+        return -1;
+    }
+
+    /** Compute partial CRC for the data input.
+     *
+     *  CRC data if not available fully, CRC can be computed in parts with available data.
+     *  Previous CRC output should be passed as argument to the current compute_partial call.
+     *  @pre: Call \ref compute_partial_start to start the partial CRC calculation.
+     *  @post: Call \ref compute_partial_stop to get the final CRC value.
+     *
+     *  @param[IN]  buffer  Data bytes
+     *  @param[IN]  size  Size of data
+     *  @param[OUT] crc  CRC value (intermediate CRC )
+     *  @return  0  on success or a negative error code on failure
+     *  @note: CRC as output in compute_partial is not final CRC value, call @ref compute_partial_stop
+     *         to get final correct CRC value.
+     */
+    virtual int32_t compute_partial(void *buffer, crc_data_size_t size, crc_size_t *crc)
+    {
+        return -1;
+    }
+
+    /** Get the final CRC value of partial computation.
+     *
+     *  CRC value available in partial computation is not correct CRC, as some
+     *  algorithms require remainder to be reflected and final value to be XORed
+     *  This API is used to perform final computation to get correct CRC value.
+     *
+     *  @param[OUT] crc  CRC result
+     */
+    virtual int32_t compute_partial_stop(crc_size_t *crc)
+    {
+        return -1;
+    }
+
+    /** Get the current CRC data size
+     *
+     * @return  CRC data size in bytes
+     */
+    uint8_t get_data_size(void) const
+    {
+        crc_width_t width = this->get_width();
+        return (width <= 8 ? 1 : (width <= 16 ? 2 : 4));
+    }
+
+    /** Get the top bit of current CRC
+     *
+     * @return  Top bit is set high for respective data width of current CRC
+     *          Top bit for CRC width less then 8 bits will be set as 8th bit.
+     */
+    uint32_t get_top_bit(void) const
+    {
+        crc_width_t width = this->get_width();
+        return (width < 8 ? (1u << 7) : (1u << (width - 1)));
+    }
+
+    /** Get the CRC data mask
```

Comment:
I'm not sure why that's necessary?

Reply:
CRC result is 32-bit, but polynomial is not always 32-bit. We need to higher mask the bits to get correct CRC in result.

Reply:
That shouldn't be part of public API then?

Reply:
These API's will be used by derived class, if not private then
1. Protected.
2. Move them to all three software implementation class.

Reply:
Protected sounds good.